File: Sean的leetcode/Leetcode/H-Index II.cpp
//https://leetcode.com/discuss/56279/concise-standard-binary-search-solution-detailed-explanation
//Time: O(logN)
class Solution {
public:
    int hIndex(vector<int>& citations) {
        int left = 0;
        int n = citations.size();
        int right = n - 1;

        while (left <= right)
        {
            int mid = (left + right) >> 1;

            //mid located at the right side.
            //H-index is located at the leftmost element of the right side.
            //Because if citation[i] >= n - i then the following is also true: citation[i+1] >= n - (i + 1)
            if (citations[mid] >= n - mid)
                right = mid - 1;
            //mid located at the left side.
            else
                left = mid + 1;
        }
        //finally left pointer is located at the leftmost element of the right side.
        return n - left;
    }
};

File: MyLeetCode/python/H-Index II.py
###
# Binary Search
# Time Complexity: O(logn) 
# Space Complexity: O(1)
###

class Solution(object):
    def hIndex(self, citations):
        """
        :type citations: List[int]
        :rtype: int
        """
        if not citations:
            return 0
        l, r = 0, len(citations) - 1
        while l+1 < r:
            mid = (l + r)/2
            r_num = len(citations) - mid 
            if citations[mid] >= r_num:
                r = mid
            else:
                l = mid
        if citations[l] >= len(citations) - l:
            return len(citations) - l
        elif citations[r] >= len(citations) - r:
            return len(citations) - r
        else:
            return 0